horreum-backend/src/main/resources/db/changeLog.xml

```diff
@@ -2277,5 +2277,70 @@
                 USING (exists (SELECT 1 FROM test WHERE test.id = test_id AND has_role2(test.owner, 'tester')));
         </sql>
     </changeSet>
+
+    <changeSet id="55" author="jwhiting">
+        <createProcedure>
+            CREATE OR REPLACE FUNCTION after_run_update_func() RETURNS TRIGGER AS $$
+            DECLARE
+                v_schema text;
+                v_schemaid integer;
+                v_has_schema boolean := false;
+            BEGIN
+                FOR v_schema IN (SELECT jsonb_path_query(NEW.data, '$.\$schema'::jsonpath)#>>'{}') LOOP
+                    v_schemaid := (SELECT id FROM schema WHERE uri = v_schema);
+                    IF v_schemaid IS NOT NULL THEN
+                        INSERT INTO run_schemas (runid, testid, prefix, uri, schemaid)
+                            VALUES (NEW.id, NEW.testid, '$', v_schema, v_schemaid);
+                        v_has_schema := true;
+                    END IF;
+                END LOOP;
+                FOR v_schema IN (SELECT jsonb_path_query(NEW.data, '$.*.\$schema'::jsonpath)#>>'{}') LOOP
+                    v_schemaid := (SELECT id FROM schema WHERE uri = v_schema);
+                    IF v_schemaid IS NOT NULL THEN
+                        INSERT INTO run_schemas (runid, testid, prefix, uri, schemaid)
+                            VALUES (NEW.id, NEW.testid, '$.*', v_schema, v_schemaid);
+                        v_has_schema := true;
+                    END IF;
+                END LOOP;
+                IF NOT v_has_schema THEN
+                    PERFORM pg_notify('calculate_tags', auth_suffix(NEW.id::text) );
+                END IF;
+                DELETE FROM dataset WHERE runid = OLD.id;
+                PERFORM pg_notify('calculate_datasets', auth_suffix(NEW.id::text) );
+                RETURN NULL;
+            END;
+            $$ LANGUAGE plpgsql;
+        </createProcedure>
+        <createProcedure>
+            CREATE OR REPLACE FUNCTION after_run_delete_func() RETURNS TRIGGER AS $$
+            BEGIN
+                DELETE FROM dataset WHERE runid = OLD.id;
+                RETURN NULL;
+            END;
+            $$ LANGUAGE plpgsql;
+        </createProcedure>
+        <sql>
+            CREATE TRIGGER after_run_delete AFTER DELETE ON run FOR EACH ROW EXECUTE FUNCTION after_run_delete_func();
+            DO $BODY$
+            DECLARE tmp_row RECORD;
+            BEGIN
+                FOR tmp_row IN (SELECT id FROM run)
+                LOOP
+                  PERFORM pg_notify('calculate_datasets', tmp_row.id || ';;tester');
```

Comment:
Setting roles to just `tester` wouldn't work - besides requiring a concrete owner role the string with roles contains the cryptographic signature - remember that we're assuming some malvolence from the `appuser` and had we allowed anyone to trigger the recalculation it would break the security model.

Another point why the migration I've asked for is not possible this way (sorry, haven't thought it all the way through).